session/session.go

```diff
@@ -164,8 +164,26 @@ func (s *session) getMembufCap() int {
 func (s *session) cleanRetryInfo() {
 	if !s.sessionVars.RetryInfo.Retrying {
 		retryInfo := s.sessionVars.RetryInfo
-		for _, stmtID := range retryInfo.DroppedPreparedStmtIDs {
-			delete(s.sessionVars.PreparedStmts, stmtID)
+		if len(retryInfo.DroppedPreparedStmtIDs) > 0 {
+			planCacheEnabled := plannercore.PreparedPlanCacheEnabled()
+			var cacheKey kvcache.Key
+			if planCacheEnabled {
+				firstStmtID := retryInfo.DroppedPreparedStmtIDs[0]
+				cacheKey = plannercore.NewPSTMTPlanCacheKey(
+					s.sessionVars, firstStmtID, s.sessionVars.PreparedStmts[firstStmtID].SchemaVersion,
+				)
+			}
+			for i, stmtID := range retryInfo.DroppedPreparedStmtIDs {
+				if planCacheEnabled {
+					if i > 0 {
+						cacheKey.(plannercore.PstmtCacheKeyMutator).SetPstmtIDSchemaVersion(
+							stmtID, s.sessionVars.PreparedStmts[stmtID].SchemaVersion,
```

Comment:
we only need to reset `stmtID`?

Reply:
for normal situation, session env are as same as last `exec` time, and here we can only got current session vars, so we CAN only reset stmtID here, it works well in most time although do nothing in some corner case(e.g. prepare exec then change sql_mode then dealloc).


--




corner case
--